samples/python/llm/utils/hook_transformers.py

```diff
@@ -0,0 +1,324 @@
+    ...     ]
+    ... )
+    >>> stopping_criteria = StoppingCriteriaList([MaxLengthCriteria(max_length=20)])
+
+    >>> outputs = model.greedy_search(
+    ...     input_ids, logits_processor=logits_processor, stopping_criteria=stopping_criteria
+    ... )
+
+    >>> tokenizer.batch_decode(outputs, skip_special_tokens=True)
+    ["It might be possible to get a better understanding of the nature of the problem, but it's not"]
+    ```"""
+    # init values
+    logits_processor = logits_processor if logits_processor is not None else LogitsProcessorList()
+    stopping_criteria = stopping_criteria if stopping_criteria is not None else StoppingCriteriaList()
+    if max_length is not None:
+        warnings.warn(
+            "`max_length` is deprecated in this function, use" " `stopping_criteria=StoppingCriteriaList([MaxLengthCriteria(max_length=max_length)])` instead.",
+            UserWarning,
+        )
+        stopping_criteria = validate_stopping_criteria(stopping_criteria, max_length)
+    pad_token_id = pad_token_id if pad_token_id is not None else self.generation_config.pad_token_id
+    eos_token_id = eos_token_id if eos_token_id is not None else self.generation_config.eos_token_id
+    if isinstance(eos_token_id, int):
+        eos_token_id = [eos_token_id]
+    eos_token_id_tensor = torch.tensor(eos_token_id).to(input_ids.device) if eos_token_id is not None else None
+    output_scores = output_scores if output_scores is not None else self.generation_config.output_scores
+    output_attentions = output_attentions if output_attentions is not None else self.generation_config.output_attentions
+    output_hidden_states = output_hidden_states if output_hidden_states is not None else self.generation_config.output_hidden_states
+    return_dict_in_generate = return_dict_in_generate if return_dict_in_generate is not None else self.generation_config.return_dict_in_generate
+
+    # init attention / hidden states / scores tuples
+    scores = () if (return_dict_in_generate and output_scores) else None
+    decoder_attentions = () if (return_dict_in_generate and output_attentions) else None
+    cross_attentions = () if (return_dict_in_generate and output_attentions) else None
+    decoder_hidden_states = () if (return_dict_in_generate and output_hidden_states) else None
+
+    # if model is an encoder-decoder, retrieve encoder attention weights and hidden states
+    if return_dict_in_generate and self.config.is_encoder_decoder:
+        encoder_attentions = model_kwargs["encoder_outputs"].get("attentions") if output_attentions else None
+        encoder_hidden_states = model_kwargs["encoder_outputs"].get("hidden_states") if output_hidden_states else None
+
+    # keep track of which sequences are already finished
+    unfinished_sequences = torch.ones(input_ids.shape[0], dtype=torch.long, device=input_ids.device)
+
+    this_peer_finished = False  # used by synced_gpus only
+    while True:
+        tic = time.time()
```

Comment:
Update all time statistic functions to be consistent with `time.perf_counter()`.